apps/src/labs/progress/ProgressContainer.tsx

```diff
@@ -0,0 +1,65 @@
+import {sendSuccessReport} from '@cdo/apps/code-studio/progressRedux';
+import {
+  getProgressLevelType,
+  ProgressLevelType,
+} from '@cdo/apps/code-studio/progressReduxSelectors';
+import ProgressManager from '@cdo/apps/labs/progress/ProgressManager';
+import {useAppDispatch} from '@cdo/apps/util/reduxHooks';
+import React, {useCallback, useEffect, useRef} from 'react';
+import {useSelector} from 'react-redux';
+import {LabState, setValidationState} from '../labRedux';
+
+interface ProgressContainerProps {
+  children: React.ReactNode;
+  appType: string;
+}
+
+/**
+ * Wrapper component for Labs that manage progress using a {@link ProgressManager}.
+ * This component manages updating the progress manager with level data, hooking
+ * its updates into Redux, and providing access for child components via context.
+ */
+const ProgressContainer: React.FunctionComponent<ProgressContainerProps> = ({
+  children,
+  appType,
+}) => {
+  const dispatch = useAppDispatch();
+  const isScriptLevel = useSelector(
+    state => getProgressLevelType(state) === ProgressLevelType.SCRIPT_LEVEL
+  );
+
+  const onProgressChange = useCallback(() => {
+    const currentState = progressManager.current.getCurrentState();
+    dispatch(setValidationState(currentState));
+
+    // Tell the external system (if there is one) about the success.
+    if (isScriptLevel && currentState.satisfied) {
+      dispatch(sendSuccessReport(appType));
+    }
+  }, [dispatch, appType, isScriptLevel]);
+
+  const progressManager = useRef<ProgressManager>(
+    new ProgressManager(onProgressChange)
+  );
+
+  const levelData = useSelector(
+    (state: {lab: LabState}) => state.lab.levelData
+  );
+
+  useEffect(() => {
+    if (levelData) {
+      progressManager.current.onLevelChange(levelData);
+    }
+  }, [levelData]);
+
+  return (
+    <ProgressManagerContext.Provider value={progressManager.current}>
+      {children}
+    </ProgressManagerContext.Provider>
```

Comment:
The model-view purist in me finds it a bit unusual to use a React class to wrap UI components to provide functionality without UI, but I guess this is the way things are done these days?

Reply:
Yeah fair, I'd say this is classifies as a [Container Component](https://www.patterns.dev/posts/presentational-container-pattern) who's job is purely to manage logic and not worry about presentation. It helps with separation of concerns so that presentational views don't need to worry about implementing this as well, and plus we can share this container with other labs.